cmd/access/node_builder/access_node_builder.go

```diff
@@ -690,7 +691,45 @@ func (builder *FlowAccessNodeBuilder) BuildExecutionSyncComponents() *FlowAccess
 					return nil, fmt.Errorf("could not check if registers db is bootstrapped: %w", err)
 				}
 
-				if !bootstrapped {
+				indexerInitialized := true
+				_, err = indexedBlockHeight.ProcessedIndex()
+				if err != nil {
+					if !errors.Is(err, storage.ErrNotFound) {
+						return nil, fmt.Errorf("could not get indexed block height: %w", err)
+					}
+					indexerInitialized = false
+				}
+
+				var registers *pStorage.Registers
+
+				if bootstrapped {
+					// if the registers db is bootstrapped AND its latest height is not the
+					// root checkpoint height, then there must be indexed data in the protocol db.
+					// otherwise, the db is likely corrupt, or the protocol db was reset but the
+					// execution-state db was kept. neither case is valid
+					registers, err = pStorage.NewRegisters(pdb)
+					if err != nil {
+						return nil, fmt.Errorf("could not create registers storage: %w", err)
+					}
+
+					// the registers db may have been bootstrapped, but the node crashed before
+					// initializing the indexer. In this case, the register db's latest height
+					// will be the root height.
+					if !indexerInitialized && registers.LatestHeight() != builder.SealedRootBlock.Header.Height {
```

Comment:
I think you are adding the check that we discussed. However, this logic is getting quite complex. And I don't quite understand why we need to check the result of `&& registers.LatestHeight() != builder.SealedRootBlock.Header.Height` .

I think it's better to be moved to a separate function/module. This piece of logic is checking whether the indexer's dependencies are in a consistent state, should not be the responsibility of the node builder. Also when this check is in the access_node_builder, it's hard to write test cases for them.

IMO, the best place for this check is in the indexer itself, because it's the indexer itself updates the states of the dependencies, it has the best knowledge of what a consistent state should be. And in our test case, we could check the dependencies must be in a consistent state after certain operations, such as finish indexing a new block. 

In other words, the constructor could take all the dependencies,  flags andconfigs, then check if dependencies are consistency, and make adjustment (force reset index), after everything is done, then return the constructed instance.